space-cli/cmd/upgrade.go

```diff
@@ -0,0 +1,246 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/sirupsen/logrus"
+	scClient "github.com/spaceuptech/space-api-go"
+	spaceApiTypes "github.com/spaceuptech/space-api-go/types"
+	"github.com/spaceuptech/space-cli/utils"
+	"github.com/txn2/txeh"
+)
+
+type resultData struct {
+	Docs []*doc `mapstructure:"result"`
+}
+
+type doc struct {
+	ID                string `mapstructure:"_id" json:"id"`
+	VersionNo         string `mapstructure:"version_no" json:"versionNo"`
+	CompatibleVersion string `mapstructure:"compatible_version" json:"compatibleVersion"`
+}
+
+// Upgrade upgrades the environment which has been setup
+func Upgrade() error {
+	const ContainerGateway string = "space-cloud-gateway"
+	const ContainerRunner string = "space-cloud-runner"
+
+	result := make(map[string]interface{})
+	if err := utils.Get(http.MethodGet, "/v1/config/env", map[string]string{}, &result); err != nil {
+		return err
+	}
+
+	currentVersion := result["version"].(string)
+	latestVersion, err := getLatestVersion(currentVersion)
+	if err != nil {
+		return err
+	}
+
+	if currentVersion == latestVersion {
+		return fmt.Errorf("current verion (%s) is up to date", currentVersion)
+	}
+
+	ctx := context.Background()
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logrus.Errorf("Unable to initialize docker client - %s", err.Error())
+		return err
+	}
+
+	// get all containers containing < space-cloud > in their name
+	args := filters.Arg("label", "app=space-cloud")
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{Filters: filters.NewArgs(args), All: true})
+	if err != nil {
+		logrus.Errorf("Unable to list containers - %s", err.Error())
+		return err
+	}
+
```

Comment:
Check if length of containers is zero. If is log and Info message (`utils.LogInfo`) that space cloud is not setup. Consider running `space-cli setup` first.